// @polkadot-api/known-chains/dist/index.mjs
import {
  chainSpec as chainSpec8
} from "./chunk-JKO5JDFG.mjs";
import {
  chainSpec as chainSpec9
} from "./chunk-QIV5AJQF.mjs";
import {
  chainSpec as chainSpec10
} from "./chunk-TP7HBWY6.mjs";
import {
  chainSpec as chainSpec11
} from "./chunk-RPHPEKN5.mjs";
import {
  chainSpec as chainSpec12
} from "./chunk-UG2BESJ2.mjs";
import {
  chainSpec as chainSpec13
} from "./chunk-LYSOHAUF.mjs";
import {
  chainSpec as chainSpec14
} from "./chunk-2FT5YKVL.mjs";
import {
  chainSpec
} from "./chunk-UTUMUPY6.mjs";
import {
  chainSpec as chainSpec2
} from "./chunk-KTAJINUE.mjs";
import {
  chainSpec as chainSpec3
} from "./chunk-JCWNBXOA.mjs";
import {
  chainSpec as chainSpec4
} from "./chunk-PJH435G5.mjs";
import {
  chainSpec as chainSpec5
} from "./chunk-NFINT5MX.mjs";
import {
  chainSpec as chainSpec6
} from "./chunk-DAB5ULPD.mjs";
import {
  chainSpec as chainSpec7
} from "./chunk-EWTURUMT.mjs";
export {
  chainSpec as ksmcc3,
  chainSpec2 as ksmcc3_asset_hub,
  chainSpec3 as ksmcc3_bridge_hub,
  chainSpec4 as polkadot,
  chainSpec5 as polkadot_asset_hub,
  chainSpec6 as polkadot_bridge_hub,
  chainSpec7 as polkadot_collectives,
  chainSpec8 as rococo_v2_2,
  chainSpec9 as rococo_v2_2_asset_hub,
  chainSpec10 as rococo_v2_2_bridge_hub,
  chainSpec11 as westend2,
  chainSpec12 as westend2_asset_hub,
  chainSpec13 as westend2_bridge_hub,
  chainSpec14 as westend2_collectives
};
